File: ljh/BAEKJOON/2579_계단오르기/ClimbStairs.js
//2579 계단 오르기
// const input = require('fs').readFileSync('/dev/stdin').toString().split('\n');
const input = require('fs').readFileSync('예제.txt').toString().split('\r\n');
const n = parseInt(input[0]);
const arr = [];

for (let i = 1; i <= n; i++) {
    arr.push(parseInt(input[i]));
}

if (n < 3) {
    switch(n) {
        case 1:
            console.log(arr[0]);
            break;
        case 2:
            console.log(arr[1] + arr[0]);
            break;
        case 3:
            console.log(arr[2] + Math.max(arr[0], arr[1]));
            break;
    }
} else {
    const f = [];
    f.push(arr[0]);
    f.push(arr[1] + arr[0]); //f(1);
    f.push(arr[2] + Math.max(arr[0], arr[1])); //f(2);
    
    for(let i = 3; i <n; i++) {
        f.push(arr[i] + Math.max(arr[i-1] + f[i-3], f[i-2]));
    }
    
    console.log(f[n-1]);
    
}

File: ljh/BAEKJOON/2579_계단오르기/ClimbStairs.java

import java.io.*;

/*
계단 오르기 (2579)
 */
public class ClimbStairs {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }

        bw.write(String.valueOf(recursion(arr, n - 1)));
        bw.flush();
    }

    public static int recursion(int[] array, int index) {
        if (index == 0)
            return array[0];

        if (index == 1)
            return array[0] + array[1];

        if (index ==2)
            return Math.max(array[0], array[1]) + array[2];

        return Math.max(recursion(array, index - 2), recursion(array, index - 3) + array[index - 1]) + array[index];
    }
}
